server/src/main/java/org/jboss/as/server/deployment/DeploymentUnitPhaseService.java

```diff
@@ -126,7 +126,8 @@ public void transition(final ServiceController<?> controller, final ServiceContr
         if (nextPhase != null) {
             final ServiceName serviceName = DeploymentUtils.getDeploymentUnitPhaseServiceName(deploymentUnit, nextPhase);
             phaseService = DeploymentUnitPhaseService.create(deploymentUnit, nextPhase);
-            phaseServiceBuilder = serviceTarget.addService(serviceName, phaseService);
+            DeploymentUnitPhaseBuilder builder = this.deploymentUnit.getAttachment(Attachments.DEPLOYMENT_UNIT_PHASE_BUILDER);
```

Comment:
Good points.  If we process the DUPs before building the service for the next phase, we only need to attach the singleton DeploymentUnitPhaseBuilder to the phase context instead of the DeploymentUnit context - which makes everything simpler.